middlewares/accesslog/logger.go

```diff
@@ -0,0 +1,147 @@
+package accesslog
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"sync/atomic"
+	"time"
+)
+
+type key int
+
+const (
+	// DataTableKey is the key within the request context used to
+	// store the Log Data Table
+	DataTableKey key = iota
+)
+
+// LogHandler will write each request and its response to the access log.
+// It gets some information from the logInfoResponseWriter set up by previous middleware.
+// Note: Current implementation collects log data but does not have the facility to
+// write anywhere.
+type LogHandler struct {
+}
+
+// NewLogHandler creates a new LogHandler
+func NewLogHandler() *LogHandler {
+	return &LogHandler{}
+}
+
+// GetLogDataTable gets the request context object that contains logging data. This accretes
+// data as the request passes through the middleware chain.
+func GetLogDataTable(req *http.Request) *LogData {
+	return req.Context().Value(DataTableKey).(*LogData)
+}
+
+func (l *LogHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request, next http.HandlerFunc) {
+	now := time.Now().UTC()
+	core := make(CoreLogData)
+
+	logDataTable := &LogData{core, req.Header, nil, nil}
+	core[StartUTC] = now
+	core[StartLocal] = now.Local()
+
+	r2 := req.WithContext(context.WithValue(req.Context(), DataTableKey, logDataTable))
+
+	var crr *captureRequestReader
+	if req.Body != nil {
+		crr = &captureRequestReader{req.Body, 0}
+		r2.Body = crr
+	}
+
+	core[RequestCount] = nextRequestCount()
+	if req.Host != "" {
+		core[RequestAddr] = req.Host
+		core[RequestHost], core[RequestPort] = silentSplitHostPort(req.Host)
+	}
+	// copy the URL without the scheme, hostname etc
+	u := &url.URL{
+		Path:       req.URL.Path,
+		RawPath:    req.URL.RawPath,
+		RawQuery:   req.URL.RawQuery,
+		ForceQuery: req.URL.ForceQuery,
+		Fragment:   req.URL.Fragment,
+	}
+	us := u.String()
+	core[RequestMethod] = req.Method
+	core[RequestPath] = us
+	core[RequestProtocol] = req.Proto
+	core[RequestLine] = fmt.Sprintf("%s %s %s", req.Method, us, req.Proto)
+
+	core[ClientAddr] = req.RemoteAddr
+	core[ClientHost], core[ClientPort] = silentSplitHostPort(req.RemoteAddr)
+	core[ClientUsername] = usernameIfPresent(req.URL)
+
+	crw := &captureResponseWriter{rw: rw}
+
+	next.ServeHTTP(crw, r2)
+
+	logDataTable.DownstreamResponse = crw.Header()
+	l.logTheRoundTrip(logDataTable, crr, crw)
+}
+
+// Close closes the Logger (i.e. the file etc).
+func (l *LogHandler) Close() error {
+	return nil
+}
+
+func silentSplitHostPort(value string) (string, string) {
```

Comment:
Can we name the return parameters `host` and `port`, respectively, for documentary reasons?